pymbar/mbar_solvers.py

```diff
@@ -289,105 +297,87 @@ def adaptive(u_kn, N_k, f_k, tol=1.0e-12, options=None):
     if tol < 1.5e-15:
         logger.info("Tolerance may be too close to machine precision to converge.")
     # keep track of Newton-Raphson and self-consistent iterations
-    nr_iter = 0
-    sci_iter = 0
+    results = jit_core_adaptive(u_kn, N_k, f_k, tol, options['gamma'], options['maximum_iterations'])
 
-    f_sci = np.zeros(len(f_k), dtype=np.float64)
-    f_nr = np.zeros(len(f_k), dtype=np.float64)
+    if options['verbose']:
+        print(f"iterations:{results['iterations']}")
+        print("gnorm_sci  gnorm_nr  choice")
+        for gsci,gnr,choice in zip(results['gnorms_sci'],results['gnorms_nr'],results['which_method']):
+            print(f"{gsci} {gnr} {choice}")
 
-    # Perform Newton-Raphson iterations (with sci computed on the way)
+    return results['f_k']
+
+def core_adaptive(u_kn, N_k, f_k, tol, gamma, maximum_iterations):
 
-    # usually calculated at the end of the loop and saved, but we need
-    # to calculate the first time.
-    g = mbar_gradient(u_kn, N_k, f_k)  # Objective function gradient.
+    nr_iter = 0
+    sci_iter = 0
 
-    for iteration in range(0, options["maximum_iterations"]):
+    which_method = []
+    gnorms_sci = []
+    gnorms_nr = []
 
+    # Perform Newton-Raphson iterations (with sci computed on the way)
+    for iteration in range(0, maximum_iterations):
```

Comment:
can you write this as jax.lax.while_loop? Just jit-compiling this for loop while be a nightmare for XLA

Reply:
> can you write this as jax.lax.while_loop?

I can take a look, though the fact that it's on the adaptive loop that is generally only called once in pymbar means it won't be as useful to do this.

Reply:
maybe it makes sense then to just jit the body of this function? (i.e. move everything inside the for loop to a separate function and just jit that)

Reply:
Ah, yes, that makes a lot of sense given the constraints. I'll try that.